### PR TITLE
ZMQ: Small cleanups in the ZMQ code

### PR DESCRIPTION
**Pull request by domob1812**
This contains various small code cleanups that make the ZMQ code easier to read and maintain (at least in my opinion).  The only functional change is that a potential memory leak is fixed that would have occured when a notifier is removed from the `notifiers` list after its callback function returned `false` (which is likely not relevant in practice but still a bug).

### CODE REVIEW

**Comment by domob1812** [comment]
I split the patch into three commits which I think make sense logically and for the review.  But of course I'm happy to squash any of the changes, or to revert parts of them (as code readability is a matter of taste).

**Comment by promag** [comment]
Concept ACK.

From a quick view LGTM.

**Comment by domob1812** [comment]
Rebased

**Comment by kostyantyn** [comment]
What do you think about wrapping the code with `zmq` namespace and removing `ZMQ/zmq` prefix from class/function names? 
Basically when the location of the file is `a/b/c.cpp` then we have:
```c++
namespace a {
namespace b {
class C {}
}
}
```

**Comment by domob1812** [comment]
@kostyantyn:  At least for now, there seems to be not much namespace usage at all in the Bitcoin code - so I'm not sure if we should change that lightly.  Also, note that `zmq` is the namespace that is also used by ZMQ's own C++ language bindings; they are currently not used, but I proposed to change that (#13957).

Also, note that C++ namespaces are [different from Java's](https://stackoverflow.com/questions/2108172/c-namespaces-comparison-to-java-packages), so that some people advise against nesting (e.g. exact correspondence to the directory structure as you suggest).

**Comment by laanwj** [comment]
Concept ACK

@jmcorgan do you mind taking a look here?

File: src/zmq/zmqabstractnotifier.h
Comment (ken2812221) [inline src/zmq/zmqabstractnotifier.h]: Missing a bunch of includes here. See appveyor CI result.

Comment (domob1812) [reply src/zmq/zmqabstractnotifier.h]: I missed that comment - I've now revised the commit to include `<string>`.  There are still CI errors, though, which I don't understand so far.  Has anyone an insight of where they come from?  I'll try to resolve them now.

File: src/zmq/zmqnotificationinterface.cpp
Comment (practicalswift) [inline src/zmq/zmqnotificationinterface.cpp]: Could use `std::transform`? :-)

File: src/zmq/zmqpublishnotifier.cpp
Comment (domob1812) [reply src/zmq/zmqpublishnotifier.cpp]: Thanks, very interesting!  From a quick look, I didn't see what my PR changes here and why it worked before - but I guess a defensive rename of the method is not a bad idea.  I did that now in a separate commit and pushed it, just to see if it fixes the CI issues.  If it does, I can include this directly in one of the commits or perhaps just keep it as it is.

Comment (domob1812) [reply src/zmq/zmqpublishnotifier.cpp]: Indeed, looks like the tests pass now.

File: src/zmq/zmqnotificationinterface.cpp
Comment (hebasto) [inline src/zmq/zmqnotificationinterface.cpp]: Why is a guard dropped?
```
#ifdef ENABLE_ZMQ
#include <zmq.h>
#endif
```

**Review by hebasto — CHANGES_REQUESTED** [review]
Concept ACK.

**Comment by domob1812** [comment]
Rebased.  Is there anything missing for me to do here?  I think this is a real improvement to the quality of the ZMQ code.

**Comment by fanquake** [comment]
@jmcorgan Are you able to take a look at this?

**Comment by domob1812** [comment]
Rebased.  Can I do anything to speed up review / merging of this one?  It should not need review by the original author or a ZMQ expert, as this is mostly just "standard" refactoring from manual memory management to the use of proper RAII (e.g. `std::unique_ptr`).

File: src/zmq/zmqabstractnotifier.h
Comment (promag) [inline src/zmq/zmqabstractnotifier.h]: d10f274a6e0291e752e25b2bd99c0472d22a420f

`MakeUnique`?

**Review by promag — COMMENTED** [review]
Code review ACK ba8b67aee28623d7fdd9b2053bc1784518a4a8ca, with some comments.

File: src/zmq/zmqnotificationinterface.cpp
Comment (domob1812) [reply src/zmq/zmqnotificationinterface.cpp]: Done.

File: src/zmq/zmqabstractnotifier.h
Comment (domob1812) [reply src/zmq/zmqabstractnotifier.h]: Done.

**Comment by domob1812** [comment]
Thanks for the review!  I've addressed your comments in follow-up commits (I do not want to force-push right now and perhaps mess up your ongoing review).  I will add them as fixups to the proper commits later when you are done with the review.

I can then also split out the creation of `TryForEachAndRemoveFailed` into a separate commit if you prefer that.

File: src/zmq/zmqnotificationinterface.cpp
Comment (domob1812) [reply src/zmq/zmqnotificationinterface.cpp]: Yes sure, if you prefer.  Let's finish the review of the code first, and then I will rework the commit history as desired afterwards (else we might mess up the ongoing review with force-pushes).

**Comment by domob1812** [comment]
Rebased

File: src/zmq/zmqnotificationinterface.cpp
Comment (hebasto) [inline src/zmq/zmqnotificationinterface.cpp:67]: 69f92ffce9f0d11d9389610e702b99d9038f3fed

If this code is touched to make it easier to read and maintain, may I suggest a simpler approach:
```
    if (!notifiers.empty()) {
        auto notificationInterface = std::unique_ptr<CZMQNotificationInterface>(new CZMQNotificationInterface());
        notificationInterface->notifiers = std::move(notifiers);

        if (notificationInterface->Initialize()) {
            return notificationInterface.release();
        }
    }

    return nullptr;
```
?

File: src/zmq/zmqpublishnotifier.cpp
Comment (hebasto) [inline src/zmq/zmqpublishnotifier.cpp:22]:
```suggestion
#include <cstdarg>
#include <cstddef>
#include <map>
#include <string>
#include <utility>
```

File: src/zmq/zmqpublishnotifier.cpp
Comment (hebasto) [inline src/zmq/zmqpublishnotifier.cpp:7]: Is this required here as no defines from `bitcoin-config.h` is used here?

File: src/zmq/zmqutil.h
Comment (hebasto) [inline src/zmq/zmqutil.h:8]: 49e2cb78563df27ff0a5d2ad985e074ee98afcce
nanonit:
```suggestion
void zmqError(const char* str);
```

File: src/zmq/zmqutil.cpp
Comment (hebasto) [inline src/zmq/zmqutil.cpp:11]: 49e2cb78563df27ff0a5d2ad985e074ee98afcce
nanonit:
```suggestion
void zmqError(const char* str)
```

**Review by hebasto — CHANGES_REQUESTED** [review]
Approach ACK 14b40f1b7ce7d0b457e163dbda4e81f5105db753

File: src/zmq/zmqpublishnotifier.cpp
Comment (domob1812) [reply src/zmq/zmqpublishnotifier.cpp:22]: Indeed, I've pushed this change together with the two nanonits below (in a single commit).

File: src/zmq/zmqpublishnotifier.cpp
Comment (domob1812) [reply src/zmq/zmqpublishnotifier.cpp]: I guess not, removed.

File: src/zmq/zmqnotificationinterface.cpp
Comment (domob1812) [reply src/zmq/zmqnotificationinterface.cpp]: Indeed, looks better.  I've updated the code accordingly (except I changed the `auto notificationInterface = ...` line slightly into what I consider even clearer; but let me know if you prefer your suggestion).

**Comment by domob1812** [comment]
Thanks for the review, @hebasto, very appreciated!  I've addressed your suggestions now with fixup commits.  I will now rebase this onto current master, and then go clean up the commit history (including splitting out the refactor as suggested by @promag and you).

**Comment by domob1812** [comment]
Rebase done, and I've now cleaned up the commit history as well (squashed in the "fixup" commits as appropriate).  Also split out the `TryForEachAndRemoveFailed` refactor into its own commit.

**Comment by hebasto** [comment]
The first commit 707c4803d5edd2d5ebd6febb33e08bb031630652 is not compiled for me.

**Comment by domob1812** [comment]
> The first commit [707c480](https://github.com/bitcoin/bitcoin/commit/707c4803d5edd2d5ebd6febb33e08bb031630652) is not compiled for me. Make iterator type `auto`?

Sorry for that, it is fixed now.  The overall PR is not changed (except I've rebased it to current `master`), but now each individual commit builds.

**Review by hebasto — CHANGES_REQUESTED** [review]
ACK 166ddcbef82ba2b6734e8f09584fb73e4f9f98a4.

Mind making the latest commit 166ddcbef82ba2b6734e8f09584fb73e4f9f98a4 a sripted-diff with the following script:
```
sed -i 's/SendMessage/SendZmqMessage/g' src/zmq/zmqpublishnotifier.*
```
?

**Comment by domob1812** [comment]
> Mind making the latest commit [166ddcb](https://github.com/bitcoin/bitcoin/commit/166ddcbef82ba2b6734e8f09584fb73e4f9f98a4) a sripted-diff with the following script:

Good point, done.

File: src/zmq/zmqnotificationinterface.cpp
Comment (instagibbs) [reply src/zmq/zmqnotificationinterface.cpp:118]: Note for reviewers: I think this is resolved?

**Comment by instagibbs** [comment]
utACK 6fe2ef2acb00b1df7f6a0c0dea1a81a1924be0e1

I restarted one travis job that appeared to be a spurious timeout.

**Review by hebasto — APPROVED** [review]
re-ACK 6fe2ef2acb00b1df7f6a0c0dea1a81a1924be0e1, only the latest commit got a scripted-diff since my [previous](https://github.com/bitcoin/bitcoin/pull/13686#pullrequestreview-487649808) review.

**Pull request merged**
